### ISSUE TITLE
EventSource ReportOutOfBandMessage message impacts program execution

### ISSUE THREAD

**Issue by sudeepagarwal** [dotnet/runtime#40799]
###  Description
   
The following is a callstack from Bing Production data center where the program would periodically freeze because EventSource's error reporting mechanism ends up indirectly calling `RtlAcquireSRWLockExclusive`
```cpp
ntdll!RtlBackoff
ntdll!RtlAcquireSRWLockExclusive
ntdll!RtlpCallVectoredHandlers
ntdll!RtlDispatchException
ntdll!RtlRaiseException
kernelbase!RaiseException
kernelbase!OutputDebugStringW
coreclr!DebugDebugger::Log
system.private.corelib!System.Diagnostics.Tracing.EventSource.ReportOutOfBandMessage(System.String)
system.private.corelib!System.Diagnostics.Tracing.EventSource.ThrowEventSourceException(System.String, System.Exception)
system.private.corelib!System.Diagnostics.Tracing.EventSource.WriteEventWithRelatedActivityIdCore(Int32, System.Guid*, Int32, EventData*)
system.private.corelib!System.Diagnostics.Tracing.EventSource.WriteEvent(Int32, Int32, System.String)
```
**This can be catastrophic to production services and we think this behavior is a poor setting in .NET Core**

### Configuration

.Net Version: **5.0.0-rc.1.20404.16**
OS Windows Server 2019, x64

**Comment by jkotas** [comment]
Should ReportOutOfBandMessage throttles the OutputDebugString to 16 events max similar to how it throttles the other places that it writes to?

**Comment by tommcdon** [comment]
@noahfalk

**Comment by noahfalk** [comment]
cc @mjsabby - I think this is the same issue you were emailing me about earlier?

> Should ReportOutOfBandMessage throttles ...

It certainly feels like a good starting point. I don't feel like I have enough understanding about the underlying issue to know if it would be sufficient.

> the program would periodically freeze ...

Can we quantify this a bit? Is the program deadlocking entirely, or does it resume execution at some point? If it resumes what is the order of magnitude of the pause (milliseconds, seconds, minutes?)

> EventSource's error reporting mechanism ends up indirectly calling RtlAcquireSRWLockExclusive ... This can be catastrophic to production services ...

kernelbase!RaiseException() usually isn't catastrophic (assuming the exception is caught as it should be here) and it isn't clear why this call path is more problematic than others. If the issue is that we are calling OutputDebugString very frequently here then @jkotas' suggestion should resolve this part of it. If the issue occurs even calling the API infrequently then we probably need more investigation.


Fwiw - From what I can see in the source nothing has changed around this behavior in .NET for a very long time. This usually means something else in the application has changed to begin provoking the issue. As of .Net Core 1.0:
[Debugger.Log() called OutputDebugString](https://github.com/dotnet/coreclr/blob/abbb8f685929c7aeaa087dae46fedc1bc2af4b17/src/vm/debugdebugger.cpp#L326)
[ReportOutOfBandMessage called Debugger.Log()](https://github.com/dotnet/coreclr/blob/abbb8f685929c7aeaa087dae46fedc1bc2af4b17/src/mscorlib/src/System/Diagnostics/Eventing/EventSource.cs#L3971)
[EventSource.ThrowEventSourceException called ReportOutOfBandMessage](https://github.com/dotnet/coreclr/blob/abbb8f685929c7aeaa087dae46fedc1bc2af4b17/src/mscorlib/src/System/Diagnostics/Eventing/EventSource.cs#L2462)
[EventSource.WriteEventWithRelativeActivityIdCore called ThrowEventSourceException](https://github.com/dotnet/coreclr/blob/abbb8f685929c7aeaa087dae46fedc1bc2af4b17/src/mscorlib/src/System/Diagnostics/Eventing/EventSource.cs#L1204)

**Comment by tommcdon** [comment]
Resolved via #40799
